src/structs/probability_tables.rs

```diff
@@ -104,112 +114,100 @@ impl ProbabilityTables {
         return num_non_zeros_context;
     }
 
-    pub fn calc_coefficient_context_7x7_aavg<const ALL_PRESENT: bool>(
+    // calculates the average of the prior values from their corresponding value in the left, above and above/left block
+    // the C++ version does one coefficient at a time, but if we do it all at the same time, the compiler vectorizes everything
+    #[inline(never)]
+    pub fn calc_coefficient_context_7x7_aavg_block<const ALL_PRESENT: bool>(
         &self,
         image_data: &BlockBasedImage,
-        coord: usize,
         block_context: &BlockContext,
-        num_nonzeros_left: u8,
-    ) -> ProbabilityTablesCoefficientContext {
-        // from compute_aavg
-        let mut best_prior: u32 = 0;
-
-        let aligned = RASTER_TO_ALIGNED[coord] as usize;
-        assert!(aligned < 64); // assert here to avoid range checks later
-
-        let log_weight = 5;
-
-        if ALL_PRESENT || self.left_present {
-            best_prior += (block_context.left(image_data).get_coefficient(aligned)).abs() as u32;
-        }
-
-        if ALL_PRESENT || self.above_present {
-            best_prior += (block_context.above(image_data).get_coefficient(aligned)).abs() as u32;
-        }
+    ) -> [i16; 49] {
+        let mut best_prior = [0; 49];
+
+        if ALL_PRESENT {
+            let left = block_context.left(image_data).get_block();
+            let above = block_context.above(image_data).get_block();
+            let above_left = block_context.above_left(image_data).get_block();
+
+            // compiler does a pretty amazing job with SSE/AVX2 here
+            for i in 0..49 {
+                // approximate average of 3 without a divide
+                // (not sure why they didn't use 22 and 10 as the coefficients since it's divided by 32, but can't change without breaking the format)
```

Comment:
Mind illuminating what the problem with the choice was? Would 11/11/10 give a better file ratio?

I was trying to weigh left pixels and above pixel at roughly twice that of above_left, but I honestly didn't try weighing them more evenly

Reply:
ok I'll change the comment that it was a deliberate choice to weight the diagonal half. Makes sense on further reflection since the diagonals would be sqrt(2) further away.